db/mongodb/mongodb.go

```diff
@@ -0,0 +1,221 @@
+		}
+		cancel()
+	}()
+
+	err = client.Ping(ctx, readpref.Primary())
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to mongodb: %w", err)
+	}
+
+	return &MongoDB{db: database, collection: collection}, nil
+}
+
+func (d *MongoDB) Close() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	return d.db.Client().Disconnect(ctx)
+}
+
+func (d *MongoDB) WriteTx() db.WriteTx {
+	return &WriteTx{
+		db:         d.db,
+		batch:      make([]mongo.WriteModel, 0),
+		inMem:      make(map[string][]byte),
+		collection: d.collection,
+	}
+}
+
+func (d *MongoDB) Get(key []byte) ([]byte, error) {
+	tx := d.WriteTx()
+	defer tx.Discard()
+	return tx.Get(key)
+}
+
+func (d *MongoDB) Iterate(prefix []byte, callback func(key, value []byte) bool) error {
+	tx := d.WriteTx()
+	defer tx.Discard()
+	return tx.Iterate(prefix, callback)
+}
+
+func (d *MongoDB) Compact() error {
+	// MongoDB handles compaction automatically, so this is a no-op.
+	return nil
+}
+
+// check that WriteTx implements the db.WriteTx interface
+var _ db.WriteTx = (*WriteTx)(nil)
+
+type WriteTx struct {
+	db         *mongo.Database
+	batch      []mongo.WriteModel
+	inMem      map[string][]byte
+	collection string
+}
+
+func (tx *WriteTx) Get(k []byte) ([]byte, error) {
+	if val, ok := tx.inMem[string(k)]; ok {
+		return val, nil
+	}
+	collection := tx.db.Collection(tx.collection)
+
+	filter := bson.M{"_id": string(k)} // Convert to string
+	var result KeyVal
+	ctx, cancel := context.WithTimeout(context.Background(), MONGODB_TIMEOUT_QUERY)
+	defer cancel()
+	err := collection.FindOne(ctx, filter).Decode(&result)
+	if err == mongo.ErrNoDocuments {
+		return nil, db.ErrKeyNotFound
+	}
+	return result.Value, err
+}
+
+func (tx *WriteTx) Iterate(prefix []byte, callback func(key, value []byte) bool) error {
+	// First, handle in-memory keys
+	for k, v := range tx.inMem {
+		if strings.HasPrefix(k, string(prefix)) {
+			if !callback([]byte(k), v) {
+				return nil
+			}
+		}
+	}
+
+	// Now, handle database keys, but skip those already handled from in-memory map
+	collection := tx.db.Collection(tx.collection)
+
+	filter := bson.M{}
+	if len(prefix) > 0 {
+		filter = bson.M{"_id": bson.M{"$regex": primitive.Regex{Pattern: "^" + string(prefix)}}}
+	}
+	cursor, err := collection.Find(context.TODO(), filter)
+	if err != nil {
+		return err
+	}
+	defer cursor.Close(context.TODO())
+
+	for cursor.Next(context.TODO()) {
+		var kv KeyVal
+		err := cursor.Decode(&kv)
+		if err != nil {
+			return err
+		}
+		if _, exists := tx.inMem[kv.Key]; exists {
+			continue
+		}
+		if !callback([]byte(kv.Key), kv.Value) {
+			break
+		}
+	}
+	return nil
+}
+
+func (tx *WriteTx) Set(k, v []byte) error {
+	update := bson.D{
+		primitive.E{Key: "$set", Value: bson.D{
+			primitive.E{Key: "value", Value: v},
+		}},
+	}
+	model := mongo.NewUpdateOneModel().SetFilter(bson.D{primitive.E{Key: "_id", Value: string(k)}}).SetUpsert(true).SetUpdate(update)
```

Comment:
the point of a WriteTx is to bulk/batch writes; why are you using NewUpdateOneModel instead of NewUpdateManyModel? note that I'm not familiar with how mongo works at all, just wondering outloud.

Reply:
So the strategy is to accumulate single model updates in the memory batch. 

Then on commit, I use collection.BulkWrite() that takes as parameter a list of models to apply. 

This is the way it usually works for mongo, at least what I found it in the official docs.